dbus-wrapper/dbus_client.c

```diff
@@ -0,0 +1,328 @@
+			break;
+		case TYPE_BOOL:
+			r = sd_bus_message_open_container(m, SD_BUS_TYPE_VARIANT, "b");
+			r = sd_bus_message_append_basic(m, 'b', &option->value.as_bool);
+			break;
+		case TYPE_INT:
+			r = sd_bus_message_open_container(m, SD_BUS_TYPE_VARIANT, "i");
+			r = sd_bus_message_append_basic(m, 'i', &option->value.as_int);
+			break;
+		default:
+			ERR("Unknown type");
+			abort();
+		}
+		r = sd_bus_message_close_container(m); /* SD_BUS_TYPE_VARIANT */
+		if (r < 0) {
+			ERR("Failed to close variant container: %s", strerror(-r));
+			goto finish;
+		}
+		r = sd_bus_message_close_container(m); /* SD_BUS_TYPE_DICT_ENTRY */
+		if (r < 0) {
+			ERR("Failed to close dict entry container: %s", strerror(-r));
+			goto finish;
+		}
+		opts = opts->next;
+	}
+	r = sd_bus_message_close_container(m); /* SD_BUS_TYPE_ARRAY */
+	if (r < 0) {
+		ERR("Failed to close array container: %s", strerror(-r));
+		goto finish;
+	}
+
+	if (ctx->argv_type == DBUS_CMD_SINGLE_ARG) {
+		r = sd_bus_message_append_basic(m, 's', ctx->argv[0] ? ctx->argv[0] : "");
+		if (r < 0) {
+			ERR("Failed to append argument: %s", strerror(-r));
+			goto finish;
+		}
+	} else if (ctx->argv_type == DBUS_CMD_MULTIPLE_ARGS) {
+		r = sd_bus_message_append_strv(m, ctx->argv);
+		if (r < 0) {
+			ERR("Failed to append arguments: %s", strerror(-r));
+			goto finish;
+		}
+	}
+
+	r = sd_bus_call(ctx->bus, m, 0, &error, &reply);
+	if (r < 0) {
+		ERR("Failed to make D-Bus call: %s (%s)",
+		    strerror(sd_bus_error_get_errno(&error)),
+		    error.message);
+		goto finish;
+	}
+
+	/* TODO: do we really need return_value? */
+	r = sd_bus_message_read(reply, "b", &return_value);
+	if (r < 0) {
+		ERR("Failed to parse response message: %s", strerror(-r));
+		goto finish;
+	}
+
+	if (!return_value) {
+		ERR("Hm... something's wrong if there's no D-Bus error set");
+	}
+
+finish:
+	sd_bus_message_unref(reply);
+	sd_bus_message_unref(m);
+	sd_bus_error_free(&error);
+	if (r < 0) {
+		sd_event *event;
+		r = sd_event_default(&event);
+		if (r < 0) {
+			ERR("Failed to get event loop: %s", strerror(-r));
+			abort();
+		}
+		sd_event_exit(event, 1);
+		sd_event_unref(event);
+	}
+	return 0;
+}
+
+static int on_command_cancel(sd_event_source *s, const struct signalfd_siginfo *si, void *userdata)
+{
+	sd_bus *bus = userdata;
+	sd_bus_error error = SD_BUS_ERROR_NULL;
+	sd_bus_message *reply;
+	int success = 1;
+	int r;
+
+	r = sd_bus_call_method(bus,
+			       "org.O1.swupdd.Client",
+			       "/org/O1/swupdd/Client",
+			       "org.O1.swupdd.Client",
+			       "Cancel",
+			       &error,
+			       &reply,
+			       "b", 0);
+	if (r < 0) {
+		ERR("Failed to call D-Bus method Cancel: %s", strerror(-r));
+		goto finish;
+	}
+
+	r = sd_bus_message_read(reply, "b", &success);
+	if (r < 0) {
+		ERR("Failed to parse reply: %s", strerror(-r));
+		goto finish;
+	}
+finish:
+	if (r < 0 || !success) {
+		ERR("Failed to cancel. Not running?");
```

Comment:
So the only purpose of this block seems like to check and see if abort() is needed because we've lost the daemon service. Could it not exit normally in this case (I guess similar for the above) with just a print statement saying it couldn't talk with the daemon? (no need for all the "failed to get event loop" messages which aren't likely helpful for end users).